04_Scheduling/03_Processes_And_Threads.md

```diff
@@ -0,0 +1,379 @@
+- *Multi Process - Multi Thread*: This is where most kernels live. It's similar to the previous case, but we can now have multiple threads per process. We won't be implementing this, but it's an easy next step.
+
+## Processes
+
+We introduced a very basic process control block in the previous chapter:
+
+```c
+typedef struct {
+    status_t status;
+    cpu_status_t context;
+} process_t;
+```
+
+While this is functional, there's a few problems:
+
+- They can't be easily identified, how do we know the difference between procesess.
+- All processes currently share the same address space, and as a byproduct, the same virtual memory allocator. 
+- We can't keep track of any resources they might be using, like file handles or network sockets.
+- We can't prioritize them, as we dont know which ones are more important.
+
+We're not going to look at how to solve all of these, but we'll the important ones.
+
+### Indentifying A Process
+
+How do we tell which process is which? We're going to use a unique number as our process id (`pid`). This will let us refer to any process by passing this number around. This is `pid` can be used for programs like `ps`, `kill` and others.
+
+While an identifier is all that's required here, it can also be nice to have a `process name`. Unlike the `pid` this isn't authoritative, it can't be used to uniquely identify a process, but it does provide a nice description.
+
+We're going to update our process struct to look like the following:
+
+```c
+typedef struct {
+    size_t pid;
+    status_t process_status;
+    cpu_status_t context;
+    char name[NAME_MAX_LEN];
+} process_t;
+```
+
+Define `NAME_MAX_LEN` to be a length of your choosing, a good starting place is 64. We've taken the approach of storing the name inside of the control block, but you could also store a pointer to a string on the heap. Using the heap would require more care when using this struct, as you'd have to be sure you managed the memory properly. 
+
+How do we assign pids? We're using to use a bump allocator: which if you'll remember is just a pointer that increases. The next section covers the details of this. It's worth noting that since we're on a 64-bit architecture and using `size_t`, we don't really have to worry about overflowing this simple allocator, as we have 18446744073709551615 possible ids. That's a lot!
+
+### Creating A New Process
+
+Creating a process is pretty trivial. We need a place to store the new `process_t` struct, in our case the static array, but you might have another data struct for it. We'll want a new function that creates a new process for us. We're going to need the starting address for the new process, and it can be nice to in an argument to this function.
```

Comment:
near the end of the second sentence: 

and it can be nice to in an argument

something missing again! :)